### PR TITLE
feat: PREVENT_DEFAULT command

### LINKED ISSUE #167: GestureHandler component blocks focus on <input/>
`GestureHandler` component blocks focus on `<input/>`. Is this an expected behavior?

---

### :+1: Without GestureHandler
```js
new DOMElement(node, {
  tagName: 'input'
});
```
without

---

### :-1: With GestureHandler
```js
new DOMElement(node, {
  tagName: 'input'
});

new GestureHandler(node).on('tap', (event) => {
  console.log(`[${event.time}] Blocked by handler!`);
});
```
with

### PR DESCRIPTION
**Pull request by alexanderGugel**
* Add `PREVENT_DEFAULT` and `PROMOTE_DEFAULT` commands
* No longer preventDefault by default
* Add `UNSUBSCRIBE` command

Fixes #167 
@DnMllr @browles  Thoughts?

### CODE REVIEW

File: dom-renderers/DOMRenderer.js [hunk 1]
```diff
@@ -100,12 +100,32 @@ function DOMRenderer (element, selector, compositor) {
...
-    // TODO preventDefault should be a separate command
+DOMRenderer.prototype.subscribe = function subscribe(type) {
     this._assertTargetLoaded();
-
-    this._target.preventDefault[type] = preventDefault;
+    this._listen(type);
     this._target.subscribe[type] = true;
+};
+
+DOMRenderer.prototype.unsubscribe = function unsubscribe(type) {
```
Comment (michaelobriena) [inline dom-renderers/DOMRenderer.js]: Needs docs

File: dom-renderers/DOMRenderer.js [hunk 2]
```diff
@@ -100,12 +100,32 @@ function DOMRenderer (element, selector, compositor) {
...
+    this._assertTargetLoaded();
+    this._listen(type);
+    this._target.subscribe[type] = false;
+};
+
+DOMRenderer.prototype.preventDefault = function preventDefault(type) {
```
Comment (michaelobriena) [inline dom-renderers/DOMRenderer.js]: Needs docs

File: dom-renderers/DOMRenderer.js [hunk 3]
```diff
@@ -100,12 +100,32 @@ function DOMRenderer (element, selector, compositor) {
...
+    this._assertTargetLoaded();
+    this._listen(type);
+    this._target.preventDefault[type] = true;
+};
+
+DOMRenderer.prototype.promoteDefault = function promoteDefault(type) {
```
Comment (michaelobriena) [inline dom-renderers/DOMRenderer.js]: needs docs

File: dom-renderers/DOMRenderer.js [hunk 4]
```diff
@@ -100,12 +100,32 @@ function DOMRenderer (element, selector, compositor) {
...
+    this._assertTargetLoaded();
+    this._listen(type);
+    this._target.preventDefault[type] = false;
+};
+
+DOMRenderer.prototype._listen = function _listen(type) {
```
Comment (michaelobriena) [inline dom-renderers/DOMRenderer.js]: Needs docs

**Comment by alexanderGugel** [comment]
@michaelobriena I mentioned the missing docs in the PR. My question was rather whether we want a preventDefault command or if we should handle that case in addUIEvent. E.g. @browles suggested having an `options` argument in `addUIEvent` instead.

Counterargument: preventDefault is DOM-specific and adding it to addUIEvent makes dynamically changing preventDefault after adding the event impossible.

Thoughts?

**Comment by michaelobriena** [comment]
I know you did, it is just a way for me to note it for myself :)

**Comment by alexanderGugel** [comment]
Ahhh ok :)

**Comment by browles** [comment]
How about just `.allowDefault`?

**Comment by alexanderGugel** [comment]
@browles :+1: changed

**Comment by FarhadG** [comment]
What's the status on this PR?

@alexanderGugel @michaelobriena

**Comment by alexanderGugel** [comment]
This should be merged ASAP IMO.

**Comment by redwoodfavorite** [comment]
Don't these conflict? https://github.com/Famous/engine/pull/215

**Comment by alexanderGugel** [comment]
Yes, they do. Why did you copy over "UNSUBSCRIBE" instead of basing off of this PR?

**Comment by redwoodfavorite** [comment]
Didn't copy anything over, just started from scratch.

We can merge this and then I can reconcile the other PR with it.

**Comment by alexanderGugel** [comment]
Sounds good. Ok.

**Comment by michaelobriena** [comment]
@alexanderGugel This is not ready to merge as there are still public functions without docs.

**Comment by alexanderGugel** [comment]
NOOOO!! Ok.

**Comment by alexanderGugel** [comment]
@michaelobriena Sorry. Thought the diff was outdated. All documented now.

**Pull request merged**